barretenberg/cpp/pil/avm/gadgets/merkle_tree.pil

```diff
@@ -0,0 +1,78 @@
+include "./poseidon2_full.pil";
+
+// Handles membership and insertion into a merkle tree
+namespace merkle_tree(256);
+    
+    pol commit sel_merkle_tree;
+    sel_merkle_tree * (1 - sel_merkle_tree) = 0;
+    // Will probably be used in permutation with other traces
+    pol commit sel_update_op;
+    sel_update_op * (1 - sel_update_op) = 0;
+    pol commit sel_membership_op;
+    sel_membership_op * (1 - sel_membership_op) = 0;
+    // Gotta stop using clk for things that are more like foreign keys
+    pol commit clk;
+    // Inputs to the gadget
+    pol commit leaf_value;
+    pol commit leaf_index;
+    pol commit path_len;
+    pol commit expected_tree_root;
+    // Output of the gadget
+    // Boolean result of the membership check
+    pol commit is_member;
+    is_member * (1 - is_member) = 0;
+
+    // These are all hinted
+    pol commit sibling_value;
+
+    // If we are not done, the path_len decrements by 1
+    sel_merkle_tree * (1 - latch) * (path_len' - path_len + 1) = 0;
+
+    pol commit latch;
+    latch * (1 - latch) = 0;
+    pol commit path_len_inv;
+    // latch == 1 when the path_len == 0
+    sel_merkle_tree * (path_len * (latch * (1 - path_len_inv) + path_len_inv) - 1 + latch) = 0;
```

Comment:
My brain is having trouble understaning this

Reply:
This looks very confusing but is our standard check for "set A = 1 when B = 0, and set A = 0 when B !=0". 
It works as follows
1) Note that `a * a_inverse == 1`, but `a_inverse` only exists if `a != 0` (0 is the only element without an inverse in the field).
2) If `latch == 0`, then the formula becomes `path_len * path_len_inv - 1 = 0` (which can only be true when `path_len != 0,` since we need a valid `path_len_inv`.
3) If `latch == 1` then the formula becomes `path_len * ( 1 - path_len_inv + path_len_inv) = 0`, simplified this is `path_len = 0`

Reply:
Ah I see.... Makes sense! I'd through this in comments (either everywhere we use it, or in one place that we can point to) :shrug: